virtual_rainforest/soil/model.py

```diff
@@ -40,57 +36,67 @@ class SoilModel(BaseModel, model_name="soil"):
     def __init__(self, update_interval: timedelta64, no_layers: int, **kwargs: Any):
 
         if no_layers < 1:
-            LOGGER.error("There has to be at least one soil layer in the soil model!")
+            log_and_raise(
+                "There has to be at least one soil layer in the soil model!",
+                InitialisationError,
+            )
         elif no_layers != int(no_layers):
-            LOGGER.error("The number of soil layers must be an integer!")
+            log_and_raise(
+                "The number of soil layers must be an integer!", InitialisationError
+            )
 
         super(SoilModel, self).__init__(update_interval, **kwargs)
         self.no_layers = int(no_layers)
         # Save variables names to be used by the __repr__
         self._repr.append("no_layers")
 
     @classmethod
-    def factory(cls, config: dict[str, Any]) -> SoilModel:
+    def factory(cls, config: dict[str, Any]) -> Optional[SoilModel]:
```

Comment:
I'm not sure I like using `None` to signal failure here. It would be clearer to have something explicit. My first thought would be to have `BaseModel.init_failed` attribute that could then be checked in a loop:

```
if any([m.init_failed for m in model_list]):
    raise_and_log("Model initialisation failure")
```

That would then have to be something that each `__init__` handles. Another option - which is probably clearer - is to leave Model `__init__` methods as things that _raise_ Exceptions and don't have to worry about setting flags. The programatic use fails at the point of failure, but the factory method traps those failure modes, converts the exception to a `log_and_raise` and returns a `FailedModel` or something like that? And then have:

```
if any([isinstance(m, FailedModel) for m in  model_list]):
        raise_and_log("Model initialisation failure")
```

That's not entirely clear - I think what I'm trying to do is to keep a separation between core functions (which fail by raising) and the wrapping for the UI, which is set up to defer failure in some locations to provide as much feedback as possible for a given stage.

Reply:
Yeah I returned a `None` in these cases because I didn't want to return a `Model` object that could be further operated on. I think I'm more inclined by the option of creating a `FailedModel` class, as it's a bit clearer. As well as being extensible for other types of model failures (e.g. failures of initialisation from data)